flaml/model.py

```diff
@@ -386,17 +393,22 @@ def on_step_end(self, args, state, control, **kwargs):
         self._init_hpo_args(kwargs)
         self._metric_name = kwargs["metric"]
 
-        X_train, y_train, X_val, y_val = self._split_train_val(
-            X_train, y_train, **kwargs
-        )
+        X_train, y_train, X_val, y_val = self._get_train_val(X_train, y_train, **kwargs)
+
         if X_train.dtypes[0] == "string":
             X_train = self._preprocess(X_train, self._task, **kwargs)
             train_dataset = Dataset.from_pandas(self._join(X_train, y_train))
-            X_val = self._preprocess(X_val, self._task, **kwargs)
-            eval_dataset = Dataset.from_pandas(self._join(X_val, y_val))
+            if X_val is not None:
+                X_val = self._preprocess(X_val, self._task, **kwargs)
+                eval_dataset = Dataset.from_pandas(self._join(X_val, y_val))
+            else:
+                eval_dataset = None
         else:
             train_dataset = Dataset.from_pandas(self._join(X_train, y_train))
-            eval_dataset = Dataset.from_pandas(self._join(X_val, y_val))
+            if X_val is not None:
+                eval_dataset = Dataset.from_pandas(self._join(X_val, y_val))
+            else:
+                eval_dataset = None

```

Comment:
Why not preprocess first and then generate the train_dataset & eval_dataset? The checking of dtypes should be in _preprocess()

Reply:
I removed the case where X_train.dtypes[0] != "string"